spdlog/Cargo.toml

```diff
@@ -37,6 +37,7 @@ native = []
 libsystemd = ["libsystemd-sys"]
 multi-thread = ["crossbeam"]
 runtime-pattern = ["spdlog-internal"]
+json-formatter = ["serde", "serde_json"]
```

Comment:
I've been thinking about this gate these days. Perhaps we can enable `JsonFormatter` when feature `serde` is enabled and feature `serde` implies dependencies `serde` and `serde_json`.

The benefit of it is that we have ability to add other serialization crates in the future, so users can choose the "backend" of `JsonFormatter` by enabling just one feature.

What do you think?

Reply:
It might add lots of dependencies for a single feature as new formatters are introduced.

Or we can just make a `SerdeFormatter` as @Lancern said to decouple the serializer from the formatter. So user can choose any serializer instead of the `serde_json`.

Reply:
> It might add lots of dependencies for a single feature as new formatters are introduced.

Hmm, makes sense. What about enable `JsonFormatter` when feature `serde_json` is enabled, so `serde_json` implies `serde`?

> Or we can just make a `SerdeFormatter` as @Lancern said to decouple the serializer from the formatter. So user can choose any serializer instead of the `serde_json`.

`SerdeFormatter` is fine, but we are in no hurry to implement it in this PR.

What I meant by my comment was that `JsonFormatter` should not be limited to only using `serde` as a backend. Semantically, `JSON` is a format and `serde` is an implementation. It seems more correct that the implementation implies the format, rather than the format implying the implementation. So that if now or in the future, there are any json crates that aren't `serde`-based, we could easily add it without breaking anything.

Reply:
`serde_json` implies `serde` sounds more reasonable.

Reply:
> So that if now or in the future, there are any json crates that aren't serde-based, we could easily add it without breaking anything.

Is this really implementable? `serde` is the de-facto standard and we already rely on the critical abstractions it provides. The "critical abstractions" I'm talking about here is the `Serialize` trait which is provided by `serde`. I cannot work out a way to bypass it and still implement a usable `JsonFormatter`.

Reply:
> Is this really implementable? `serde` is the de-facto standard [...] I cannot work out a way to bypass it and still implement a usable `JsonFormatter`.

It is. For now, there are crates [`rkyv`](https://crates.io/crates/rkyv), [`miniserde`](https://crates.io/crates/miniserde), [`nanoserde`](https://crates.io/crates/nanoserde), etc., they each have their own advantages and are not `serde`-based. If we can harmlessly preserve the maximum possibilities, why not do so?

> [...] and we already rely on the critical abstractions it provides. The "critical abstractions" I'm talking about here is the `Serialize` trait which is provided by `serde`.

If you are talking about deriving `serde::Serialize` macro for types of `spdlog-rs`, this can be unforced. Considering the following psuode-code:

```rust
#[cfg(and(feature = "serde", feature = "rkyv"))
compile_error!("you can only select one serialization backend");

#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[cfg_attr(feature = "rkyv", derive(rkyv::Serialize))]
pub enum Level {
    // ...
}
```

Reply:
@SpriteOvO OK this makes sense. I thought you mean we completely abandon any dependencies on serialization libraries and let the users choose what suits them. So we're still depending on a bunch of pre-selected serialization libraries (even if they're just optional) and let the user choose one among them.